File: app/src/main/java/com/dcrandroid/activities/ReaderActivity.java
/*
 * Copyright (c) 2018-2019 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.activities;

import android.Manifest;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

import com.dcrandroid.R;
import com.dcrandroid.data.Constants;
import com.google.zxing.integration.android.IntentIntegrator;
import com.google.zxing.integration.android.IntentResult;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

/**
 * Created by Macsleven on 11/15/2015.
 */

public class ReaderActivity extends AppCompatActivity implements ActivityCompat.OnRequestPermissionsResultCallback {

    String address;

    /**
     * Called when the activity is first created.
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            View decorView = getWindow().getDecorView();
            decorView.setSystemUiVisibility(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS |
                    View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR);
        }
        setContentView(R.layout.custom_ui_read);

        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
            requestCameraPermission();
        } else {
            startCamera();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent intent) {
        if (requestCode == IntentIntegrator.REQUEST_CODE) {
            IntentResult scanResult = IntentIntegrator.parseActivityResult(requestCode, resultCode, intent);
            if (scanResult.getContents() != null) {
                address = scanResult.getContents();
                intent.putExtra(Constants.ADDRESS, address);
                setResult(RESULT_OK, intent);
                finish();
            } else {
                finish();
            }
        }
    }

    private void requestCameraPermission() {
        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
            int allowed = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA);
            if (allowed == PackageManager.PERMISSION_DENIED) {
                new AlertDialog.Builder(this).setTitle(R.string.permission)
                        .setMessage(R.string.camera_permission_scan)
                        .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                String[] permissions = {Manifest.permission.CAMERA, Manifest.permission.VIBRATE};
                                ActivityCompat.requestPermissions(ReaderActivity.this, permissions, 200);
                            }
                        }).setCancelable(false)
                        .show();
            } else {
                startCamera();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == 200) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //permission granted
                startCamera();
            } else {
                Toast.makeText(this, R.string.denied_permission, Toast.LENGTH_SHORT).show();
                finish();
            }
        }
    }

    private void startCamera() {
        IntentIntegrator integrator = new IntentIntegrator(this);
        integrator.setCaptureActivity(CaptureActivityAnyOrientation.class);
        integrator.setOrientationLocked(false);
        integrator.setDesiredBarcodeFormats(IntentIntegrator.QR_CODE);
        integrator.setPrompt(getString(R.string.scan_info));
        integrator.setCameraId(0);  // Use a specific camera of the device
        integrator.setBeepEnabled(false);
        integrator.setBarcodeImageEnabled(false);
        integrator.initiateScan();
    }

}

File: app/src/main/java/com/dcrandroid/activities/ReaderActivity.kt
/*
 * Copyright (c) 2018-2021 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.activities

import android.Manifest
import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.widget.Toast
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.dcrandroid.R
import com.dcrandroid.data.Constants
import com.dcrandroid.dialog.InfoDialog
import com.google.zxing.BarcodeFormat
import com.google.zxing.ResultPoint
import com.journeyapps.barcodescanner.BarcodeCallback
import com.journeyapps.barcodescanner.BarcodeResult
import com.journeyapps.barcodescanner.DefaultDecoderFactory
import kotlinx.android.synthetic.main.qr_reader_layout.*

class ReaderActivity : BaseActivity(), ActivityCompat.OnRequestPermissionsResultCallback, BarcodeCallback {

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.qr_reader_layout)

        val allowed = ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
        if (allowed == PackageManager.PERMISSION_DENIED) {
            val dialog = InfoDialog(this)
                    .setDialogTitle(getString(R.string.permission))
                    .setMessage(getString(R.string.camera_permission_scan))
                    .setPositiveButton(getString(R.string.ok), DialogInterface.OnClickListener { _, _ ->
                        val permissions = arrayOf(Manifest.permission.CAMERA, Manifest.permission.VIBRATE)
                        ActivityCompat.requestPermissions(this@ReaderActivity, permissions, 200)
                    })
            dialog.setCancelable(false)
            dialog.show()
        } else {
            startCamera()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == 200) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                //permission granted
                startCamera()
            } else {
                Toast.makeText(this, R.string.denied_permission, Toast.LENGTH_SHORT).show()
                finish()
            }
        }
    }

    override fun onPause() {
        super.onPause()
        barcode_scanner.pause()
    }

    override fun onResume() {
        super.onResume()
        barcode_scanner.resume()
    }

    override fun barcodeResult(result: BarcodeResult?) {
        if (result != null) {
            val data = Intent()
            data.putExtra(Constants.RESULT, result.text)
            setResult(Activity.RESULT_OK, data)
            finish()
        }
    }

    override fun possibleResultPoints(resultPoints: MutableList<ResultPoint>?) {}

    private fun startCamera() {
        val formats = listOf(BarcodeFormat.QR_CODE)
        barcode_scanner.barcodeView.decoderFactory = DefaultDecoderFactory(formats)
        barcode_scanner.decodeSingle(this)
        barcode_scanner.setStatusText(getString(R.string.qr_code_scan_hint))
        barcode_scanner.initializeFromIntent(intent)
    }

}
